### DISCUSSION TITLE
[paused] stage 1: Auth / RBAC

### DISCUSSION THREAD

**Discussion by atherdon** [atherdon/dash-back#11] [General]
tags & keywords это скорее фича, чем чтото важное сейчас. но т.к там ничего сложного тоже нет - можно и их перенести. просто в тот момент, когда мы сделаем деплой и девы смогут фетчить данные - нам tags\keywords погоды сейчас не делают. тоже самое и с progress.

---

нужно понять как перекинуть первые данные фронтенд девелоперам, чтобы они могли убить у себя ту папку data. заодно они могут написать список с calcualtions, которые нам чуть позже понадобится сделать, чтобы данные просто статикой выводились внутри таблиц.

--- 

### stage 1: authentification: to be discussed - simple for 10 ppl or RBAC with 2-3 roles

> Наши основные разработчики, которые делают весь проект - немного скептически относятся к тому, что я посути решил убить их код и начать все с нуля. причина простая - они поддерживают сложный проект и на изменения админки у них нету времени. плюс админка это скучно. Поэтому когда я им начал показывать наш код - они сразу смогли меня уколоть, сказав что проект сейчас не очень секьюрный. с чем я согласен - но это не была основная проблема.
> С другой стороны - раз уж мы начали разбивать проект, которому 2 месяца уже на две части - может нам стоит сразу предусмотреть auth и rbac. 
> 

единственная проблема - я ненавижу эти формы логинов и т.д. поэтому я думал найти какой то простой вариант, чтобы можно было прикрутить какой то oauth0 например и забыть об этом.

с другой стороны - если делать нормальный RBAC - скорее всего кастомный вариант лучше. посути у нас не будет много польвателей, кому нужен доступ к админке. ну 8 редакторов. ну допустим еще сделаем каждому разрабочику по аккаунут чтобы они могли тестить. в общем вопрос интересный, но сейчас он в low priority - потому что отдельный backend отодвигает запуск этого проекта немного назад.

**Comment by kolserdav** [comment]
Для фронтенда достаточно файлов `docs/API.md` и постман коллекции в `resources/postman`.

**Reply by atherdon** [reply]
ок, тогда я передам это @vadim9999 и посмотрим как произойдет connection.

**Reply by atherdon** [reply]
Давайте тогда обсудим  как мы должны сделать auth. поидее, оба варианта(simple auth & rbac) уже реализовывали на graphql и не раз. поэтому мы можем попробывать или адаптировать чужой код или сделать свою версию.
я конечно за то чтобы оптимизировать чужую схему - чтобы было проще.

для текущей версии проекта - auth который позволит работать внутри 10 людям - полностью удовлетворяет текущий scope. с другой стороны - тк мы начинаем с чистого листа - возможно расширенная версия не займет так много времени как мне кажется.

нужно ваше мнение @kolserdav 

ссылки которые я быстро нагуглил
- https://graphql.org/learn/authorization/
- https://github.com/kkemple/graphql-auth

rbac
- https://medium.com/swlh/rule-based-authorisation-for-your-graphql-server-cf81bfa60790
- https://github.com/Canner/graphql-rbac
- https://auth0.com/blog/dynamic-authorization-with-graphql-and-rules/

**Reply by kolserdav** [reply]
>     * https://github.com/kkemple/graphql-auth

В версии `3.3.0` использовал опыт создания хука из https://github.com/kkemple/graphql-auth, отказавшись от небезопасного посредника на уровне `express`. @atherdon спасибо)

Теперь можно закрывать роуты:
- по минимальной роли
- по определённым ролям
- разрешать доступ к `своим` материалам независимо от двух предыщих ограничений
- запрещать изменение определенных полей, при доступе по предыдущему пункту, если один из первых двух не разрешает

Управляется просто, при объявлении резолвера используется хук `withAuth`, код для добавления дополнительных правил доступа к своим материалам вынесен в отдельный файл `graphq/checkSelfSchema.ts`

**Reply by atherdon** [reply]
Вынес этот комментарий в отдельный таск - потом когда будем роли подключать в dashb - сразу понадобится.

нужно ли мне сделать список ролей, чтобы вы их залили в базу сразу? я могу сделать обычный js array и добавить его в папку data - на будущее cc @kolserdav

**Reply by kolserdav** [reply]
Сейчас роли это просто integer где 0 - простой юзер. Ограничений нет по количеству ролей. Можно и именовать их, но тогда нужна их нумерация для закрытия по минимальной роли, если оно планируется на будущее.

**Reply by atherdon** [reply]
Ok, я подумаю об этом. приятно что ограничений нет и можно придумать все что хочется.
изза того что прошла дашбоард не могла растягиваться - приходилось и приходится постоянно чтото выдумывать

**Comment by kolserdav** [comment]
Как реализовать посредника это детали. Главный вопрос как хранить сессии. Знаю два варианта.
1. Без хранения в базе. `jsonwebtoken` - шифрует по секретному ключу ид пользователя, и часть хеша пароля (чтобы блокировать если пароль был изменен).
2. Auth token - уникальная строка хранящаяся в базе или в Redis.
В первом случае можно сделать постоянный токен так как ответственность за его сохранность лежит на самом пользователе, а во втором естесственно только по времени жизни токена и каждый запрос обновлять время.

**Reply by kolserdav** [reply]
Первый тоже можно сделать временным, если создавать вместе с refresh токеном, в случае просрочки перевыпускать токен. То есть если Auth просрочен смотрит Auth-refresh если нету или не верифицирован отказ.

**Reply by kolserdav** [reply]
Вопроса нет, какой токен выбрать, конечно jwt, вопрос - вечный или по времени, если по времени, то какая политика в этом отношении?

**Reply by atherdon** [reply]
у меня возникла такая идея: бекенд у нас сейчас полностью independent. фронтенд будет к нему подключаться "извне" и тоже хранится в отдельном репозитории.
Что если, мы будем делать auth на бекенде, а параллельно например сделаем чистый реакт проект, задача которого будет только в том - чтобы сделать авторизацию на фронтенде и все.

таким макаром мы можем убить несколько зайцев:
- мы можем делать api на бекенде и сразу вы сможете его тестировать на фронте
- не нужно будет много обсуждений с текущей фронтенд командой
- их часть проекта не будет стопорится, и они просто продолжат делать views для вывода данных

**Reply by atherdon** [reply]
> Вопроса нет, какой токен выбрать, конечно jwt, вопрос - вечный или по времени, если по времени, то какая политика в этом отношении?

мне кажется можно и вечный делать сейчас. потому что пользователей то будет немного у этого проекта.
это не какой то сверх нагруженный Sass и т.д. но думаю что вам впринципе виднее. как скажете так наверное и будет

**Reply by kolserdav** [reply]
> у меня возникла такая идея: бекенд у нас сейчас полностью independent. фронтенд будет к нему подключаться "извне" и тоже хранится в отдельном репозитории.
> Что если, мы будем делать auth на бекенде, а параллельно например сделаем чистый реакт проект, задача которого будет только в том - чтобы сделать авторизацию на фронтенде и все.

Интрересная идея, но какая цель?
Тут два минуса:
1. Тогда не будет реактивной навигации между авторизацией и самим приложением.
2. Невозможность на фронте в переиспользовании компонентов, т.е. те же кнопки и поля будут отдельно описаны в auth и в приле.

Зато этот фронт сразу будет как Oauth сервис, потому что после успешной авторизации будет редиректить с токеном на целевую страницу приложения. Таким образом, в будущем его можно будет использовать и для других сервисов, где подойдёт единый аккаунт. Если такая цель есть - тогда супер если нет, тогда не понятно, что мы получаем,  взамен минусов.

**Reply by atherdon** [reply]
просто я хотел еще предложить похожую логику например для upload данных в базу.
т.к. эта часть будет нужна только в первые пару месяцев, пока мы не сделаем sync с основной базой - мой помощник редактора будет забивать данные руками. 

и чтобы не городить все в одном фронтенде - я тоже думал предложить сделать отдельный фронт(еще один) в котором просто будут формы, которые будут автоматически сгенерированны реактом на основе graphql схемы

главный плюс в том, что вы сможете быть лидером как в создании backend части авторизации, так и фронтенд части. еще один плюс, что мы сможем например выложить альфа версию без авторизации - что будет быстрее, а потом уже сможем прикрутить вашу логику

**Reply by atherdon** [reply]
Предлагаю разбить stage 1 на части, чтобы нам проще было с оценкой и быстрее стартануть.

первая часть: auth c jwt - делаете как считаете будет лучше. я пытался подумать пару дней как делать лучше, но ничего не придумал. Плюс, нехочу чтобы вы ждали. если rbac это просто - вообще круто, можем тогда это сразу добавить - лишним точно не будет.

вторая часть: чистый react app только для auth который делаете вы и коннектим его к graphql server. 

сейчас нужен эстимейт по бекенд части, чтобы я сразу запросил бюджет и чтобы не ждать. Для прошлого жстимейта я просто заранее брал деньги.

**Reply by atherdon** [reply]
я могу рассказать детали, а вы уже примете решение: 

1. Нам важно делать быстро проект, поэтому я даже бекенд вначале не хотел создать 
2. Думаю что для начала мне на секьюрити наплевать, потому что пока мы используем старую дашбоард - компания просто теряет больше денег. 

3. Возможно нам лучше сделать импорт а потом думать про аутентификацию. 
4. Если вы смогли затащить ещё мелкий проект - это бы сьекономило мне время и я смог бы сделать нормальное описание какие схемы нам нужно сделать.

5. Проект наш будет переделываться несколько раз, текущая версия не должна быть идеальной. Потому что через месяца 2-3 нам придется коннектится к основному ядру проекта, а какой там код внутри я не знаю

**Reply by kolserdav** [reply]
Получается, нужно сделать, регистрацию, аутентификацию, и встроить https://github.com/kkemple/graphql-auth или найти другое решение для авторизации, и закрыть несколько узлов для примера?

**Reply by atherdon** [reply]
> Получается, нужно сделать, регистрацию, аутентификацию, и встроить https://github.com/kkemple/graphql-auth или найти другое решение для авторизации, и закрыть несколько узлов для примера?

Думаю да. мне кажется этого нам хватит. А если уже возникнет вопрос в том что нужно расширить функционал - у prisma хорошо сделан процесс апдейтов схем.

**Reply by atherdon** [reply]
@kolserdav я на всякий создам проект по stage 1 react-auth, а условия мы всегда сможем изменить. просто нехочу вас тормозить со своей стороны

**Reply by atherdon** [reply]
не удержался

![P4Z0pudvoUUIUu72EhuivHLqxJp2-ql1c3377](https://user-images.githubusercontent.com/1469198/107735634-d96e9100-6d08-11eb-9570-d4a6de903abc.jpeg)

**Comment by kolserdav** [comment]
"можно было прикрутить какой то oauth0 например и забыть об этом"

Oauth это если у вас есть сервер с реализованной авторизацией, то можно перенаправлять туда там авторизуется оттудв присылается временный одноразовый токен по которому можно получить его публичные данные. На сервере нужно чтобы проверял document.referer при заходе на страницу авторизации и query string с ид запроса  и в случае успеха если всё сходится перенаправлять на страницу с токеном и ид запроса в uri.

Если рассматривать Oauth популярных сервисов, то нужно чтобы сервис соответствовал их требованиям в плане бизнес правил интернета.

**Comment by kolserdav** [comment]
"с другой стороны - если делать нормальный RBAC "

Если auth реализован, rbac это просто название роли в профиле поьзователя и описание их уровней плюс один if в посреднике, который сверяет минимально допустимую роль контекста с ролью пользователя. Или  точное совпадение, если допуск к контексту не по range а по equal или enum.

**Comment by kolserdav** [comment]
По авторизации, сложности испытал с графкл(, не стал костыли лепить, лишь бы сделать, на завтра перенес, подумать надо.

**Reply by atherdon** [reply]
это хорошо что застряли немного - значит начинается интересная часть

**Comment by kolserdav** [comment]
У меня кончается триал на auth0, с формами закончу, продолжу с ним разбираться, уже в рамках фри аккаунта.

**Comment by atherdon** [comment]
я пока бы поставил даже не начинал эту часть делать. мне очень лень.

можете сделать delete account - потому что я хочу найти какое то элегантное
решение для этой ситуации. но пока его не вижу.
сейчас у нас просто на фронте стоит эта хрень и даже ничего не отдает. пока
работает.

>

**Reply by kolserdav** [reply]
Вот я там тоже тогда сходу не смог данные запрашивать, авторизуется и всё. А данные нужны, чтобы идентифицировать пользователя и назначить ему роль. И получается всё равно придётся хранить емайл, иначе по какому принципу назначать кастомные роли?! Если это будет просто идентификаторы, то админу всё равно придётся по ним спрашивать емайлы, чтобы знать кому какую роль.

**Reply by atherdon** [reply]
проблема тут еще в том, что я пока незнаю, хочу ли я внутри нашей админки хрнаить имейлы. вариант который сейчас - работает - хоть и странно. но если мы захотим сделать sync с основной базой - чтобы имейлы и пароли от старой админки подходили тут - прийдется пилить костыли. мы попробывали сделать быструю атаку на эту проблему - не получилось.
значит откатываемся назад пока что. просто может через пару месяцев это даже и не будет проблемой

**Reply by kolserdav** [reply]
Админка написана в stage1, там только пару секьюрных вещей добавить: регистрация только по приглашению и антиспам при входе. А так там уже всё по тем ролям можно перекрывать. Получается она не подходит?

**Reply by kolserdav** [reply]
По сбору емайлов сделать галку о принятии политики конфиденциальности а по хранению, вынести юзера в отдельную базу а пароль и пользователя зашифровать по ключу, который на сервере нигде не записан, а запускать с этим ключом в переменной среды.

**Reply by kolserdav** [reply]
Ну и супер юзера базы пользователя никому не давать)

**Reply by kolserdav** [reply]
И root сервера и sudo не давать чужим, иначе может просто отключить аутентификацию сервера базы и залезть куда не следует)

**Reply by kolserdav** [reply]
Если всё это сделать, то цена получения этих емайлов будет в тысячи раз больше их номинальной стоимости.

**Reply by kolserdav** [reply]
Я не хочу вас торопить, просто хотел убедится, что Вы помните про stage1.

**Reply by atherdon** [reply]
я пока еще не придумал как решить этот вопрос :)

моя секьюрити сейчас находится на этапе отправления .env файлов в телеграме разным девам. поэтому мне сложно сейчас придумать как сделать auth так - как надо. вот решим еще проблем - я обещаю подумать как сделать.

**Reply by kolserdav** [reply]
Не вопрос:) Просто, возможно мои наводящие вопросы смогут вам пригодиться.
Когда будете по этому вопросу принимать решение, предлагаю сначала попробовать набросать какие запросы кому будут разрешены, а какие кому нет. А от этой политики уже отталкиваться на каком инструменте реализовывать.

**Comment by atherdon** [comment]
Я как раз очень рад тому что мы обсуждаем эти stages. потому что в итоге
получается даже лучше чем было задумано.

**Comment by atherdon** [comment]
Я все надеюсь на то что когда мы выкатим крутые изменения в Апреле -
основные девы охренеют и может сами напилят auth :)
Но они оч заняты. поэтому я только им рассказываю, но ничего не показываю
